library/core/src/str/count.rs

```diff
@@ -0,0 +1,116 @@
+        let mut counts = 0;
+        let unrolled_chunks = chunk.array_chunks::<UNROLL_INNER>();
+        // If there's a remainder (know can only happen for the last item in
+        // `chunks`, because `CHUNK_SIZE % UNROLL == 0`), then we need to
+        // account for that (although we don't use it to later).
+        let remainder = unrolled_chunks.remainder();
+        for unrolled in unrolled_chunks {
+            for &word in unrolled {
+                // Because `CHUNK_SIZE` is < 256, this addition can't cause the
+                // count in any of the bytes to overflow into a subsequent byte.
+                counts += contains_non_continuation_byte(word);
+            }
+        }
+
+        // Sum the values in `counts` (which, again, is conceptually a `[u8;
+        // size_of::<usize>()]`), and accumulate the result into `total`.
+        total += sum_bytes_in_usize(counts);
+
+        // If there's any data in `remainder`, then handle it. This will only
+        // happen for the last `chunk` in `body.chunks()` (because `CHUNK_SIZE`
+        // is divisible by `UNROLL_INNER`), so we explicitly break at the end
+        // (which seems to help LLVM out).
+        if !remainder.is_empty() {
+            // Accumulate all the data in the remainder.
+            let mut counts = 0;
+            for &word in remainder {
+                counts += contains_non_continuation_byte(word);
+            }
+            total += sum_bytes_in_usize(counts);
+            break;
+        }
+    }
+    total
+}
+
+// Checks each byte of `w` to see if it contains the first byte in a UTF-8
+// sequence. Bytes in `w` which are continuation bytes are left as `0x00` (e.g.
+// false), and bytes which are non-continuation bytes are left as `0x01` (e.g.
+// true)
+#[inline]
+fn contains_non_continuation_byte(w: usize) -> usize {
+    let lsb = 0x0101_0101_0101_0101u64 as usize;
+    ((!w >> 7) | (w >> 6)) & lsb
+}
+
+// Morally equivalent to `values.to_ne_bytes().into_iter().sum::<usize>()`, but
+// more efficient.
+#[inline]
+fn sum_bytes_in_usize(values: usize) -> usize {
+    const LSB_SHORTS: usize = 0x0001_0001_0001_0001_u64 as usize;
+    const SKIP_BYTES: usize = 0x00ff_00ff_00ff_00ff_u64 as usize;
+
+    let pair_sum: usize = (values & SKIP_BYTES) + ((values >> 8) & SKIP_BYTES);
+    pair_sum.wrapping_mul(LSB_SHORTS) >> ((core::mem::size_of::<usize>() - 2) * 8)
+}
+
+// This is the most direct implementation of the concept of "count the number of
+// bytes in the string which are not continuation bytes", and is used for the
+// head and tail of the input string (the first and last item in the tuple
+// returned by `slice::align_to`).
+fn char_count_general_case(s: &[u8]) -> usize {
+    const CONT_MASK_U8: u8 = 0b0011_1111;
+    const TAG_CONT_U8: u8 = 0b1000_0000;
+    let mut leads = 0;
+    for &byte in s {
+        let is_lead = (byte & !CONT_MASK_U8) != TAG_CONT_U8;
```

Comment:
No benefit from `(byte as i8) >= -64` here? In the old implementation it was slightly faster, I think because it could directly use the result from the flag register. This should be mostly relevant in very short strings.

Reply:
Hm, I'll double check. At one point it mattered (it was more that LLVM happened to unroll this function less aggressively than the others, since it can't really tell that the input should be small).

That said, this code has changed a lot since then.

Reply:
I've gone back to the old impl. It seems very slightly faster, although these measurements are so noisy :(.